gobblin-core/src/main/java/org/apache/gobblin/source/extractor/extract/QueryBasedSource.java

```diff
@@ -241,7 +244,13 @@ public int hashCode() {
       addLineageSourceInfo(state, sourceEntity, workunit);
       partition.serialize(workunit);
       workUnits.add(workunit);
+      highestWaterMark = Math.max(highestWaterMark, partition.getHighWatermark());
+      lowestWaterMark = Math.min(lowestWaterMark, partition.getLowWatermark());
     }
+    state.appendToListProp(TimingEvent.FlowEventConstants.HIGH_WATERMARK_FIELD,
+        String.format("%s.%s: %s", sourceEntity.getDatasetName(), sourceEntity.destTableName, highestWaterMark));
+    state.appendToListProp(TimingEvent.FlowEventConstants.LOW_WATERMARK_FIELD,
+        String.format("%s.%s: %s", sourceEntity.getDatasetName(), sourceEntity.destTableName, lowestWaterMark));
```

Comment:
could `partitions` ever be empty?  if so, I recommend lowest, highest to be `Optional` and only write conditionally, rather than setting the properties to `MAX_VALUE` and `-1`.